File: wallendar-app/app/src/main/java/com/td/wallendar/addgroup/ui/AddGroupPresenter.java
package com.td.wallendar.addgroup.ui;

import com.td.wallendar.dtos.request.AddGroupRequest;
import com.td.wallendar.models.Group;
import com.td.wallendar.repositories.interfaces.GroupsRepository;
import com.td.wallendar.utils.scheduler.SchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.disposables.CompositeDisposable;

public class AddGroupPresenter {

    private final WeakReference<AddGroupView> view;
    private final GroupsRepository groupsRepository;
    private final CompositeDisposable disposable;
    private final SchedulerProvider schedulerProvider;


    public AddGroupPresenter(final AddGroupView addGroupView, final GroupsRepository groupsRepository, final SchedulerProvider schedulerProvider) {
        view = new WeakReference<>(addGroupView);
        this.groupsRepository = groupsRepository;
        this.disposable = new CompositeDisposable();
        this.schedulerProvider = schedulerProvider;
    }

    public void createGroup(final String groupTitle, final Long ownerId) {
        disposable.add(groupsRepository.createGroup(new AddGroupRequest(groupTitle, ownerId))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onGroupCreatedSuccessfully, this::OnGroupCreatedWithError));
    }

    private void onGroupCreatedSuccessfully(final Group group) {
        view.get().onGroupCreated(group);
    }

    private void OnGroupCreatedWithError(Throwable throwable) {
        //TODO error
    }

    public void onViewDetached() {
        disposable.dispose();
    }
}

File: wallendar-app/app/src/main/java/com/td/wallendar/addgroup/ui/AddGroupPresenter.kt
package com.td.wallendar.addgroup.ui

import com.td.wallendar.dtos.request.AddEventRequest
import com.td.wallendar.dtos.request.AddGroupRequest
import com.td.wallendar.models.Event
import com.td.wallendar.models.Group
import com.td.wallendar.repositories.interfaces.EventsRepository
import com.td.wallendar.repositories.interfaces.GroupsRepository
import com.td.wallendar.utils.scheduler.SchedulerProvider
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference
import java.text.SimpleDateFormat
import java.util.*

class AddGroupPresenter(
        addGroupView: AddGroupView,
        private val groupsRepository: GroupsRepository,
        private val eventsRepository: EventsRepository,
        private val schedulerProvider: SchedulerProvider,
        private val isEvent: Boolean,
) {
    private val view: WeakReference<AddGroupView> = WeakReference(addGroupView)
    private val disposable: CompositeDisposable = CompositeDisposable()

    fun createGroup(groupTitle: String, ownerId: Long) {
        disposable.add(groupsRepository.createGroup(AddGroupRequest(groupTitle, ownerId))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ group: Group -> onGroupCreatedSuccessfully(group) }) { throwable: Throwable -> onGroupCreatedWithError(throwable) })
    }

    fun createEvent(eventTitle: String, ownerId: Long, date: Calendar) {
        val dateFormatted = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss").format(date.time)
        disposable.add(eventsRepository.createEvent(AddEventRequest(eventTitle, ownerId, dateFormatted))
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ event: Event -> onGroupCreatedSuccessfully(event) }) { throwable: Throwable -> onGroupCreatedWithError(throwable) })
    }

    private fun onGroupCreatedSuccessfully(group: Group) {
        view.get()?.onGroupCreated(group)
    }

    private fun onGroupCreatedWithError(throwable: Throwable) {
        view.get()?.onGroupCreatedWithErrors()
    }

    fun onViewDetached() {
        disposable.dispose()
    }

}